mlir/lib/Analysis/DataFlowFramework.cpp

```diff
@@ -67,8 +67,10 @@ void ProgramPoint::print(raw_ostream &os) const {
        << OpWithFlags(getPrevOp(), OpPrintingFlags().skipRegions());
     return;
   }
-  os << "<before operation>:"
-     << OpWithFlags(getNextOp(), OpPrintingFlags().skipRegions());
+  if (!isBlockEnd()) {
```

Comment:
When can this happen? Is the block empty? In your example, I don't see an empty block. Is there something else that we can print here?

Reply:
A program point is basically a `Block::iterator`. There are two program points inside of the `scf.forall`: before the memref.load and after the memref.load. Both cases should be supported by the current implementation. Only case I can think of is "empty block", but that's not the case here. Am I missing something?

Reply:
Sorry, I forgot: The scf.forall.in_parallel is a designated terminator for the scf.forall operation.You can see https://mlir.llvm.org/docs/Dialects/SCFDialect/#scfforallin_parallel-scfinparallelop.